conduit/plugins/exporters/filewriter/util.go

```diff
@@ -1,52 +1,76 @@
+// EncodingFormat enumerates the acceptable encoding formats for Conduit file-based plugins.
+type EncodingFormat byte
+
+const (
+	// MessagepackFormat indicates the file is encoded using MessagePack.
+	MessagepackFormat EncodingFormat = iota
+
+	// JSONFormat indicates the file is encoded using JSON.
+	JSONFormat
+
+	// UnrecognizedFormat indicates the file's encoding is unknown to Conduit.
+	UnrecognizedFormat
+)
+
+var jsonPrettyHandle *codec.JsonHandle
 
 func init() {
-	prettyHandle = new(codec.JsonHandle)
-	prettyHandle.ErrorIfNoField = json.CodecHandle.ErrorIfNoField
-	prettyHandle.ErrorIfNoArrayExpand = json.CodecHandle.ErrorIfNoArrayExpand
-	prettyHandle.Canonical = json.CodecHandle.Canonical
-	prettyHandle.RecursiveEmptyCheck = json.CodecHandle.RecursiveEmptyCheck
-	prettyHandle.Indent = json.CodecHandle.Indent
-	prettyHandle.HTMLCharsAsIs = json.CodecHandle.HTMLCharsAsIs
-	prettyHandle.MapKeyAsString = true
-	prettyHandle.Indent = 2
-
-	jsonStrictHandle = new(codec.JsonHandle)
-	jsonStrictHandle.ErrorIfNoField = prettyHandle.ErrorIfNoField
-	jsonStrictHandle.ErrorIfNoArrayExpand = prettyHandle.ErrorIfNoArrayExpand
-	jsonStrictHandle.Canonical = prettyHandle.Canonical
-	jsonStrictHandle.RecursiveEmptyCheck = prettyHandle.RecursiveEmptyCheck
-	jsonStrictHandle.Indent = prettyHandle.Indent
-	jsonStrictHandle.HTMLCharsAsIs = prettyHandle.HTMLCharsAsIs
-	jsonStrictHandle.MapKeyAsString = true
+	jsonPrettyHandle = new(codec.JsonHandle)
+	jsonPrettyHandle.ErrorIfNoField = json.CodecHandle.ErrorIfNoField
+	jsonPrettyHandle.ErrorIfNoArrayExpand = json.CodecHandle.ErrorIfNoArrayExpand
+	jsonPrettyHandle.Canonical = json.CodecHandle.Canonical
+	jsonPrettyHandle.RecursiveEmptyCheck = json.CodecHandle.RecursiveEmptyCheck
+	jsonPrettyHandle.Indent = json.CodecHandle.Indent
+	jsonPrettyHandle.HTMLCharsAsIs = json.CodecHandle.HTMLCharsAsIs
+	jsonPrettyHandle.MapKeyAsString = true
+	jsonPrettyHandle.Indent = 2
 }
 
-// EncodeJSONToFile is used to encode an object to a file. If the file ends in .gz it will be gzipped.
-func EncodeJSONToFile(filename string, v interface{}, pretty bool) error {
-	var writer io.Writer
+// ParseFilenamePattern parses a filename pattern into an EncodingFormat and gzip flag.
+func ParseFilenamePattern(pattern string) (EncodingFormat, bool, error) {
+	originalPattern := pattern
+	gzip := false
+	if strings.HasSuffix(pattern, ".gz") {
+		gzip = true
+		pattern = pattern[:len(pattern)-3]
+	}
+
+	var blockFormat EncodingFormat
+	if strings.HasSuffix(pattern, ".msgp") {
+		blockFormat = MessagepackFormat
+	} else if strings.HasSuffix(pattern, ".json") {
+		blockFormat = JSONFormat
+	} else {
+		return UnrecognizedFormat, false, fmt.Errorf("unrecognized export format: %s", originalPattern)
+	}
+
+	return blockFormat, gzip, nil
+}
 
+// EncodeToFile enocods an object to a file using a given format and possible gzip compression.
+func EncodeToFile(filename string, v interface{}, format EncodingFormat, isGzip bool) error {
```

Comment:
can the format be inferred from the filename extension?

Reply:
Here I'm proposing to separate the responsibility of deciding which encoding from which filename extension. The exporter's `Init()` parses the configured (or default) file pattern and sets its `format` field. So the caller of `EncodeToFile()` has to ensure that it's using the appropriate extension for a given format.

For this exporter's `Receive()`, the filename is constructed appropriately via:
```go
path.Join(exp.cfg.BlocksDir, fmt.Sprintf(exp.cfg.FilenamePattern, exportData.Round()))
```